### ISSUE TITLE
odd handling of whitespace characters

### ISSUE THREAD

**Issue by Fishrock123** [npm/npm-package-arg#35]
It seems that for tags / tags with versions, whitespace characters cannot be part of the "name", but _are_ allowed as part of the "version"?

This very surprising and confusing.

I think that npa should just trim all whitespace out since it cannot exist in npm package names or semver versions anyways (to my knowledge), and it would make it easier for some tools to accept inputs such as `"name @ version"`.

```
npa("hello@ 1.0.0")
{ type: 'version',
  registry: true,
  where: undefined,
  raw: 'hello@ 1.0.0',
  name: 'hello',
  escapedName: 'hello',
  scope: undefined,
  rawSpec: ' 1.0.0',
  saveSpec: null,
  fetchSpec: ' 1.0.0',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined }
```

```
npa("hello @1.0.0")
Thrown:
{ Error: Invalid package name "hello ": name cannot contain leading or trailing spaces; name can only contain URL-friendly characters
    at invalidPackageName (/Users/Jeremiah/Documents/here-be-dragons/node_modules/npm-package-arg/npa.js:82:15)
    at Result.setName (/Users/Jeremiah/Documents/here-be-dragons/node_modules/npm-package-arg/npa.js:118:11)
    at new Result (/Users/Jeremiah/Documents/here-be-dragons/node_modules/npm-package-arg/npa.js:107:23)
    at resolve (/Users/Jeremiah/Documents/here-be-dragons/node_modules/npm-package-arg/npa.js:54:15)
    at npa (/Users/Jeremiah/Documents/here-be-dragons/node_modules/npm-package-arg/npa.js:48:10) code: 'EINVALIDPACKAGENAME' }
```

**Comment by isaacs** [comment]
But `" 1.0.0"` _is_ a valid semver _range_, because whitespace characters are logical `AND`, and an empty string is equivalent to `*`.

```
> new semver.Range(' 1.1.1 ')
Range {
  options: {},
  loose: false,
  includePrerelease: false,
  raw: ' 1.1.1 ',
  set: [ [ [Comparator] ] ],
  range: '1.1.1'
}
```

Since "anything and also 1.1.1" resolves to just "1.1.1", the semver.Range class strips these out of the trimmed down result to save on verification, but it is technically valid and npa is doing the right thing.

```
> npa('hello@ 1.1.1 ')
Result {
  type: 'range',
  registry: true,
  where: undefined,
  raw: 'hello@ 1.1.1 ',
  name: 'hello',
  escapedName: 'hello',
  scope: undefined,
  rawSpec: ' 1.1.1 ',
  saveSpec: null,
  fetchSpec: ' 1.1.1 ',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined
}```

**Comment by isaacs** [comment]
Oh, wait, I spoke too soon.  If the space comes before the version, but not after it, then you get this strange result: 

```
> npa('hello@ 1.1.1')
Result {
  type: 'version',
  registry: true,
  where: undefined,
  raw: 'hello@ 1.1.1',
  name: 'hello',
  escapedName: 'hello',
  scope: undefined,
  rawSpec: ' 1.1.1',
  saveSpec: null,
  fetchSpec: ' 1.1.1',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined
}
```
